src/app/consensus_test/consensus_test.ml

```diff
@@ -0,0 +1,229 @@
+  match select_outcome with
+  | CandidateLonger ->
+      candidate_block :: !current_chain
+  | EqualLength -> (
+      match !current_chain with
+      | _ :: rest_of_list ->
+          candidate_block :: rest_of_list
+      | [] ->
+          !current_chain )
+  | CandidateShorter ->
+      !current_chain
+
+let run_select ~context:(module Context : CONTEXT)
+    (existing_block : Mina_block.Precomputed.t)
+    (candidate_block : Mina_block.Precomputed.t) =
+  let existing_consensus_state_with_hashes =
+    { With_hash.hash =
+        Mina_state.Protocol_state.hashes existing_block.protocol_state
+    ; data =
+        Mina_state.Protocol_state.consensus_state existing_block.protocol_state
+    }
+  in
+  let candidate_consensus_state_with_hashes =
+    { With_hash.hash =
+        Mina_state.Protocol_state.hashes candidate_block.protocol_state
+    ; data =
+        Mina_state.Protocol_state.consensus_state candidate_block.protocol_state
+    }
+  in
+  match
+    Consensus.Hooks.select
+      ~context:(module Context)
+      ~existing:existing_consensus_state_with_hashes
+      ~candidate:candidate_consensus_state_with_hashes
+  with
+  | `Take ->
+      let candidate_length =
+        Mina_state.Protocol_state.consensus_state candidate_block.protocol_state
+        |> Consensus.Data.Consensus_state.blockchain_length
+        |> Unsigned.UInt32.to_int
+      in
+      let existing_length =
+        Mina_state.Protocol_state.consensus_state existing_block.protocol_state
+        |> Consensus.Data.Consensus_state.blockchain_length
+        |> Unsigned.UInt32.to_int
+      in
+      compare_lengths candidate_length existing_length
+  | `Keep ->
+      CandidateShorter
+
+let process_precomputed_blocks ~context ~current_chain blocks =
+  let%bind () =
+    Deferred.List.iter blocks ~f:(fun candidate_block ->
+        let existing_block = List.hd_exn !current_chain in
+        let select_outcome =
+          run_select ~context existing_block candidate_block
+        in
+        current_chain :=
+          update_chain ~current_chain ~candidate_block ~select_outcome ;
+        return () )
+  in
+  return ()
+
+let write_blocks_to_output_dir ~current_chain ~output_dir =
+  let sorted_output =
+    List.map ~f:precomputed_block_to_block_file_output !current_chain
+    |> List.rev
+  in
+  let write_block_to_file (i : int) (block : BlockFileOutput.t) :
+      unit Deferred.t =
+    let block_json_str =
+      block |> BlockFileOutput.to_yojson |> Yojson.Safe.to_string
+    in
+    let output_file = sprintf "%s/block_%d.json" output_dir i in
+    Writer.save output_file ~contents:block_json_str
+  in
+  let%bind () =
+    Deferred.List.iteri sorted_output ~f:(fun i block ->
+        let%bind () = write_block_to_file i block in
+        return () )
+  in
+  return ()
+
+let main () ~blocks_dir ~output_dir =
+  let logger = Logger.create () in
+  let current_chain : Mina_block.Precomputed.t list ref = ref [] in
```

Comment:
NIT: this reference seems to not be necessary. A more idiomatic way of implementing this code in OCaml would be to have `process_precomputed_blocks` return a list instead of writing into one. This can be done easily by using `Deferred.List.fold_left` instead of `Deferred.List.iter`.